packages/components/src/services/hds-theming.ts

```diff
@@ -0,0 +1,233 @@
+import Service from '@ember/service';
+import { tracked } from '@glimmer/tracking';
+
+export enum HdsThemeValues {
+  // system settings (prefers-color-scheme)
+  System = 'system',
+  // user settings for dark/light
+  Light = 'light',
+  Dark = 'dark',
+}
+
+enum HdsModesBaseValues {
+  Hds = 'hds', // TODO understand if it should be `default`
+}
+
+enum HdsModesLightValues {
+  CdsG0 = 'cds-g0',
+  CdsG10 = 'cds-g10',
+}
+
+enum HdsModesDarkValues {
+  CdsG90 = 'cds-g90',
+  CdsG100 = 'cds-g100',
+}
+
+export enum HdsCssSelectorsValues {
+  Data = 'data',
+  Class = 'class',
+}
+
+export type HdsThemes = `${HdsThemeValues}`;
+export type HdsModes =
+  | `${HdsModesBaseValues}`
+  | `${HdsModesLightValues}`
+  | `${HdsModesDarkValues}`
+  | undefined;
+export type HdsModesLight = `${HdsModesLightValues}`;
+export type HdsModesDark = `${HdsModesDarkValues}`;
+export type HdsCssSelectors = `${HdsCssSelectorsValues}`;
+
+type HdsThemingOptions = {
+  lightTheme: HdsModesLight;
+  darkTheme: HdsModesDark;
+  cssSelector: HdsCssSelectors;
+};
+
+type SetThemeArgs = {
+  theme: HdsThemes | undefined;
+  options?: HdsThemingOptions;
+  onSetTheme?: OnSetThemeCallback;
+};
+
+export type OnSetThemeCallbackArgs = {
+  currentTheme: HdsThemes | undefined;
+  currentMode: HdsModes | undefined;
+};
+
+export type OnSetThemeCallback = (args: OnSetThemeCallbackArgs) => void;
+
+export const THEMES: HdsThemes[] = Object.values(HdsThemeValues);
+export const MODES_LIGHT: HdsModesLight[] = Object.values(HdsModesLightValues);
+export const MODES_DARK: HdsModesDark[] = Object.values(HdsModesDarkValues);
+export const MODES: HdsModes[] = [
+  ...Object.values(HdsModesBaseValues),
+  ...MODES_LIGHT,
+  ...MODES_DARK,
+];
+
+export const HDS_THEMING_DATA_SELECTOR = 'data-hds-theme';
+export const HDS_THEMING_CLASS_SELECTOR_PREFIX = 'hds-theme';
+export const HDS_THEMING_CLASS_SELECTORS_LIST = [
+  ...MODES_LIGHT,
+  ...MODES_DARK,
+].map((mode) => `${HDS_THEMING_CLASS_SELECTOR_PREFIX}-${mode}`);
+
+export const HDS_THEMING_LOCALSTORAGE_DATA = 'hds-theming-data';
+
+export const DEFAULT_THEMING_OPTION_LIGHT_THEME = HdsModesLightValues.CdsG0;
+export const DEFAULT_THEMING_OPTION_DARK_THEME = HdsModesDarkValues.CdsG100;
+export const DEFAULT_THEMING_OPTION_CSS_SELECTOR = 'data';
+
+export default class HdsThemingService extends Service {
+  @tracked _isInitialized: boolean = false;
+  @tracked _currentTheme: HdsThemes | undefined = undefined;
+  @tracked _currentMode: HdsModes = undefined;
```

Comment:
Would it be clearer to include "undefined" for the type here (vs. including it in the type definition above) the same way it's defined for `_currentTheme` or is that a dumb question? (Or change the type definition for `HdsThemes`?)

(i.e. use `HdsModes | undefined = undefined;`)

Reply:
I had a few difficulties including `undefined` in the `HdsThemes` type definition, like I've done for the `HdsModes`. TypeScript was complaining in some places, so I ended up using a slightly different definition.

Reply:
I think I may understand what you meant. I've made the update here: https://github.com/hashicorp/design-system/pull/3240/commits/fd95fc1107d9db650ad1cc8ccf90f658a308ad4a